### PR TITLE
download: limit the number of retries

### PR DESCRIPTION
**Pull request by pjonsson**
The token will eventually expire, so
stop trying after 5 attempts failed.
This will result in a failure downloading
instead of a hanging job, so the root
cause remains, but the impact is reduced.

Refs #84

### CODE REVIEW

**Comment by pjonsson** [comment]
Beyond limiting the number of attempts to 5, this PR preserves the existing logic and just moves code around which is a deliberate choice to lower the risk of introducing bugs. The choice of 5 attempts is arbitrary, it's much less than an infinite number of attempts, but more than immediately giving up.

I don't think the current values correspond to exactly what I will need in a month or two, but if anyone needs to fine-tune these values that can easily be done in a followup PR. As an added bonus, a PR that is just about fine-tuning values will be easy to review.

**Comment by pedep** [comment]
Code looks fine as-is, but i will have to test it before merging.
The retry you refactored is not only used to catch 401 errors, but also the 429 errors returned by the error-prone rate-limiter used by copernicus.

Since its now possible for the retry request loop to give up, we should implement a way for the user to collect the errored. This could be a "resume file", or simply returning the ids of failed objects instead of `None`

**Comment by pedep** [comment]
Could you deliberate on what you mean by this?
> moves code around which is a deliberate choice to lower the risk of introducing bugs.

**Comment by pjonsson** [comment]
> The retry you refactored is not only used to catch 401 errors, but also the 429 errors returned by the error-prone rate-limiter used by copernicus.

Can you rephrase what you wrote? The expired key is never renewed in the current code, so after getting a 401, the download is stuck in that loop forever?

> Could you deliberate on what you mean by this?
>
> >  moves code around which is a deliberate choice to lower the risk of introducing bugs.

I resisted the temptation to try to make everything perfect, this PR is just a transplant of what is currently in the code, besides limiting the number of attempts.

**Comment by pjonsson** [comment]
I'm not using the cli, but I've used this code for a couple of days and it does back out of the infinite loop when getting a 401 response.

I have never gotten a 429 that I'm aware of, but #99 should improve the situation with that since the Retry class respects Retry-After with its default parameters to the constructor. (I was not aware there was a rate limiter, so if the rate limiter is completely broken #99 will obviously not help.)

**Comment by pedep** [comment]
> The retry you refactored is not only used to catch 401 errors, but also the 429 errors returned by the error-prone rate-limiter used by copernicus.

The download endpoint may return 429 errors in some edge cases. You can reproduce this by force restarting a multi-threaded download a few times in the span of one minute. The rate limiter will still have memory of your previous connections, and return 429 to your current connections. If you increase the concurrency above your account limit (regular accounts have a limit of 4) you can force the issue.

> I resisted the temptation to try to make everything perfect, this PR is just a transplant of what is currently in the code, besides limiting the number of attempts.

Thats fair.
I can see you also changed enough to fix the issue mentioned here https://github.com/SDFIdk/CDSETool/issues/84#issuecomment-1992083460 :+1:

**Review by pedep — APPROVED** [review]
I will refrain from requesting the change to the return value, as i can see we also return `None` in a few other places.

**Comment by pjonsson** [comment]
> I will refrain from requesting the change to the return value, as i can see we also return `None` in a few other places.

Thanks! I'll mention that returning an ID when the download fails will unfix #59, but 
I'm guessing `download_features` has enough information to correlate what was passed in to `download_feature` and what was returned.
